_posts/2017-05-30-predicting-artist-career-stage-with-machine-learning-in-apache-spark.markdown

```diff
@@ -0,0 +1,181 @@
+
+val trainedModel = model.run(labeledPointsForTraining)
+
+logInfo(s"weights: ${trainedModel.weights}, intercept: ${trainedModel.intercept}")
+```
+
+We can use the generated model to run predictions on known data and compare results using `RegressionMetrics`.
+
+```scala
+import org.apache.spark.mllib.evaluation._
+
+val predictionsAndObservations = labeledPointsForTraining.map { point =>
+  val prediction = model.predict(point.features)
+  (prediction, point.label)
+}
+
+val metrics = new RegressionMetrics(predictionsAndObservations)
+
+logInfo(s"mean squared error: ${metrics.meanSquaredError}")
+logInfo(s"r2=${metrics.r2}")
+logInfo(s"explained variance=${metrics.explainedVariance}")
+```
+
+The lower the MSE, the better and should be used to tune the optimizer above.
+
+### Predicting Career Stage
+
+Same operation, but using artists that don't have a career stage value.
+
+```scala
+val featuresForPredicting = features.filter(feature => feature.careerStageGene == 0)
+
+val results = featuresForPredicting.map { feature =>
+  val predictedCareerStage = trainedModel.predict(scaler.transform(feature.vector))
+  (feature.artistId, predictedCareerStage)
+}
+```
+
+### Tests
+
+After refactoring the above functionality into `train` and `predict` functions we can write some tests.
+
+```scala
+import net.artsy.utils.SparkSpec
+import org.apache.spark.mllib.feature.StandardScaler
+import org.apache.spark.mllib.regression.LabeledPoint
+
+"train" should "train a model with intercept" in {
+  val featuresForTraining = Array(
+    Features(artistId = "1", careerStageGene = 50.0, age = 50, maxPrice = 10.0),
+    Features(artistId = "2", careerStageGene = 60.0, age = 60, maxPrice = 20.0),
+    Features(artistId = "3", careerStageGene = 40.0, age = 0, maxPrice = 0)
+  )
+
+  val scaler = new StandardScaler().fit(sc.parallelize(featuresForTraining.map(f => f.vector)))
+  val labeledPointsForTraining = featuresForTraining.map {
+    f => LabeledPoint(f.careerStageGene, scaler.transform(f.vector))
+  }
+  val trainedModel = train(sc.parallelize(labeledPointsForTraining))
+
+  trainedModel.weights.numActives should equal(2)
+  trainedModel.intercept should equal(40.0 +- 1.5)
+}
+
+"predict" should "predict career stage genes" in {
+  val features = Array(
+    Features(artistId = "1", careerStageGene = 50.0, age = 50, maxPrice = 10.0),
+    Features(artistId = "2", careerStageGene = 60.0, age = 60, maxPrice = 20.0),
+    Features(artistId = "3", careerStageGene = 40.0, age = 0, maxPrice = 0),
+    Features(artistId = "4", careerStageGene = 0.0, age = 55, maxPrice = 15.0),
+    Features(artistId = "5", careerStageGene = 0.0, age = 35, maxPrice = 75.0)
+  )
+
+  val predictions = predict(sc.parallelize(features)).collect().toMap
+  predictions.get("4").get should equal(55.0 +- 1)
+  predictions.get("5").get should equal(58.5 +- 1)
+}
+```
+
+### Conclusion
+
+Machine learning is really easy in Spark! Use the above example as a cookbook.
```

Comment:
Sounds trite. Maybe you can comment about all the number-crunching is outsourced to libraries, or link to our past ML posts.